### ISSUE TITLE
Switching to Nvidia doesn't turn off Intel graphics

### ISSUE THREAD

**Issue by Alarg** [Askannz/optimus-manager#291]
I've done a fresh reinstall of Manjaro system. I tried to install the program, yet I encounter this error.
When I'm on Intel graphics, I get this output for `xrandr --listproviders`:
```
[alarg@ALEKSA ~]$ xrandr --listproviders
Providers: number : 1
Provider 0: id: 0x44 cap: 0xf, Source Output, Sink Output, Source Offload, Sink Offload crtcs: 3 outputs: 2 associated providers: 0 name:modesetting
```
When I switch to Nvidia, I get the following:
```
[alarg@ALEKSA ~]$ xrandr --listproviders
Providers: number : 2
Provider 0: id: 0x1b8 cap: 0x1, Source Output crtcs: 0 outputs: 0 associated providers: 1 name:NVIDIA-0
Provider 1: id: 0x1dd cap: 0xf, Source Output, Sink Output, Source Offload, Sink Offload crtcs: 3 outputs: 2 associated providers: 1 name:modesetting

```
I'm not sure where the problem comes from, logs show that everything operates correctly yet the iGPU doesn't switch off. What should I check to make this go?

Hybrid mode:
```
[alarg@ALEKSA ~]$ xrandr --listproviders
Providers: number : 2
Provider 0: id: 0x44 cap: 0xf, Source Output, Sink Output, Source Offload, Sink Offload crtcs: 3 outputs: 2 associated providers: 0 name:modesetting
Provider 1: id: 0x23f cap: 0x0 crtcs: 0 outputs: 0 associated providers: 0 name:NVIDIA-G0
```

**Comment by Alarg** [comment]
Contents of the log file:
Switching from Intel to Nvidia:
```
[6] INFO: # Xorg pre-start hook
[6] INFO: Previous state was: {'type': 'pending_pre_xorg_start', 'requested_mode': 'nvidia', 'current_mode': 'intel'}
[6] INFO: Requested mode is: nvidia
[6] INFO: Checking for GDM display servers
[200] INFO: Available modules: ['nouveau', 'bbswitch', 'acpi_call', 'nvidia', 'nvidia_drm', 'nvidia_modeset', 'nvidia_uvm']
[200] INFO: Unloading modules ['nouveau'] (if loaded)
[203] INFO: Loading module bbswitch
[205] INFO: Setting GPU power to ON via bbswitch
[361] INFO: Setting Nvidia PCI power state to on
[401] INFO: Loading module nvidia
[618] INFO: Loading module nvidia_drm
[1288] INFO: Loaded extra Intel Xorg options (0 lines)
[1288] INFO: Loaded extra Intel Xorg options (0 lines)
[1289] INFO: Writing to /etc/X11/xorg.conf.d/10-optimus-manager.conf
[1289] INFO: Writing state {'type': 'pending_post_xorg_start', 'switch_id': '20200713T141750', 'requested_mode': 'nvidia'}
[1289] INFO: Xorg pre-start hook completed successfully.
[3] INFO: # Xorg post-start hook
[3] INFO: Running xrandr commands
[269] INFO: Running /etc/optimus-manager/xsetup-nvidia.sh
[280] INFO: Writing state {'type': 'done', 'switch_id': '20200713T141750', 'current_mode': 'nvidia'}
[280] INFO: Xorg post-start hook completed successfully.

```
Switching from Nvidia to Hybrid:
```
[6] INFO: # Xorg pre-start hook
[6] INFO: Previous state was: {'type': 'pending_pre_xorg_start', 'requested_mode': 'hybrid', 'current_mode': 'nvidia'}
[6] INFO: Requested mode is: hybrid
[6] INFO: Checking for GDM display servers
[31] INFO: Loaded extra Intel Xorg options (0 lines)
[31] INFO: Loaded extra Intel Xorg options (0 lines)
[32] INFO: Writing to /etc/X11/xorg.conf.d/10-optimus-manager.conf
[32] INFO: Writing state {'type': 'pending_post_xorg_start', 'switch_id': '20200713T141915', 'requested_mode': 'hybrid'}
[32] INFO: Xorg pre-start hook completed successfully.
[3] INFO: # Xorg post-start hook
[3] INFO: Running /etc/optimus-manager/xsetup-hybrid.sh
[18] INFO: Writing state {'type': 'done', 'switch_id': '20200713T141915', 'current_mode': 'hybrid'}
[19] INFO: Xorg post-start hook completed successfully.
```
Switching from Hybrid to Intel:
```
[6] INFO: # Xorg pre-start hook
[6] INFO: Previous state was: {'type': 'pending_pre_xorg_start', 'requested_mode': 'intel', 'current_mode': 'hybrid'}
[6] INFO: Requested mode is: intel
[6] INFO: Checking for GDM display servers
[193] INFO: Available modules: ['nouveau', 'bbswitch', 'acpi_call', 'nvidia', 'nvidia_drm', 'nvidia_modeset', 'nvidia_uvm']
[193] INFO: Unloading modules ['nvidia_drm', 'nvidia_modeset', 'nvidia_uvm', 'nvidia'] (if loaded)
[664] INFO: Loading module bbswitch
[668] INFO: Setting GPU power to OFF via bbswitch
[741] WARNING: bbswitch is enabled, pci_power_control option ignored.
[766] INFO: Loaded extra Intel Xorg options (0 lines)
[767] INFO: Loaded extra Intel Xorg options (0 lines)
[767] INFO: Writing to /etc/X11/xorg.conf.d/10-optimus-manager.conf
[767] INFO: Writing state {'type': 'pending_post_xorg_start', 'switch_id': '20200713T142201', 'requested_mode': 'intel'}
[767] INFO: Xorg pre-start hook completed successfully.
[3] INFO: # Xorg post-start hook
[3] INFO: Running /etc/optimus-manager/xsetup-intel.sh
[10] INFO: Writing state {'type': 'done', 'switch_id': '20200713T142201', 'current_mode': 'intel'}
[10] INFO: Xorg post-start hook completed successfully.

```

**Comment by Alarg** [comment]
/var/log/daemon logs:
Intel -> Nvidia:
```
[24] INFO: # Daemon pre-start hook
[24] INFO: Removing /etc/X11/xorg.conf.d/10-optimus-manager.conf (if present)
[25] INFO: Copying /etc/optimus-manager/optimus-manager.conf to /var/lib/optimus-manager/tmp/config_copy.conf
[27] INFO: Startup mode is: intel
[27] INFO: Writing state {'type': 'pending_pre_xorg_start', 'requested_mode': 'intel', 'current_mode': None}
[27] INFO: Daemon pre-start hook completed successfully.
[27] INFO: Calling Xorg pre-start hook.
[1] INFO: # Commands daemon
[1] INFO: Opening UNIX socket
[1] INFO: Awaiting commands
[98538] INFO: Received command : {
    "args": {
        "content": "[intel]\nDRI=3\naccel=\ndriver=modesetting\nmodeset=yes\ntearfree=\n\n[nvidia]\nDPI=96\nPAT=yes\nallow_external_gpus=no\nignore_abi=yes\nmodeset=yes\noptions=overclocking\n\n[optimus]\nauto_logout=yes\npci_power_control=yes\npci_remove=no\npci_reset=function_level\nstartup_auto_battery_mode=intel\nstartup_auto_extpower_mode=nvidia\nstartup_mode=intel\nswitching=bbswitch\n"
    },
    "type": "user_config"
}

[98539] INFO: Replacing user config at /etc/optimus-manager/optimus-manager.conf with provided content
[98540] INFO: Received command : {
    "args": {
        "path": ""
    },
    "type": "temp_config"
}

[98540] INFO: Removing temporary config file path
[120974] INFO: Received command : {
    "args": {
        "mode": "nvidia"
    },
    "type": "switch"
}

[120974] INFO: Writing requested GPU mode nvidia
[120974] INFO: Writing state {'type': 'pending_pre_xorg_start', 'requested_mode': 'nvidia', 'current_mode': 'intel'}
[164858] INFO: Received command : {
    "args": {
        "mode": "intel"
    },
    "type": "switch"
}

[164858] INFO: Writing requested GPU mode intel
[164859] INFO: Writing state {'type': 'pending_pre_xorg_start', 'requested_mode': 'intel', 'current_mode': 'nvidia'}
[434673] INFO: Process stop requested
[434673] INFO: Closing and removing the socket...
[434673] INFO: Goodbye !

```
Nvidia -> Hybrid:
```
[29] INFO: # Daemon pre-start hook
[29] INFO: Removing /etc/X11/xorg.conf.d/10-optimus-manager.conf (if present)
[30] INFO: Copying /etc/optimus-manager/optimus-manager.conf to /var/lib/optimus-manager/tmp/config_copy.conf
[33] INFO: Startup mode is: intel
[33] INFO: Writing state {'type': 'pending_pre_xorg_start', 'requested_mode': 'intel', 'current_mode': None}
[33] INFO: Daemon pre-start hook completed successfully.
[33] INFO: Calling Xorg pre-start hook.
[1] INFO: # Commands daemon
[1] INFO: Opening UNIX socket
[1] INFO: Awaiting commands
[2575043] INFO: Process stop requested
[2575043] INFO: Closing and removing the socket...
[2575043] INFO: Goodbye !

```
Hybrid -> Intel:
```
[29] INFO: # Daemon pre-start hook
[30] INFO: Removing /etc/X11/xorg.conf.d/10-optimus-manager.conf (if present)
[31] INFO: Copying /etc/optimus-manager/optimus-manager.conf to /var/lib/optimus-manager/tmp/config_copy.conf
[35] INFO: Startup mode is: intel
[36] INFO: Writing state {'type': 'pending_pre_xorg_start', 'requested_mode': 'intel', 'current_mode': None}
[36] INFO: Daemon pre-start hook completed successfully.
[36] INFO: Calling Xorg pre-start hook.
[2] INFO: # Commands daemon
[2] INFO: Opening UNIX socket
[2] INFO: Awaiting commands
[62030] INFO: Received command : {
    "args": {
        "content": "[intel]\nDRI=3\naccel=\ndriver=modesetting\nmodeset=yes\ntearfree=\n\n[nvidia]\nDPI=96\nPAT=yes\nallow_external_gpus=no\nignore_abi=yes\nmodeset=yes\noptions=overclocking\n\n[optimus]\nauto_logout=yes\npci_power_control=yes\npci_remove=no\npci_reset=no\nstartup_auto_battery_mode=intel\nstartup_auto_extpower_mode=nvidia\nstartup_mode=intel\nswitching=bbswitch\n"
    },
    "type": "user_config"
}

[62031] INFO: Replacing user config at /etc/optimus-manager/optimus-manager.conf with provided content
[62032] INFO: Received command : {
    "args": {
        "path": ""
    },
    "type": "temp_config"
}

[62032] INFO: Removing temporary config file path
[65291] INFO: Received command : {
    "args": {
        "mode": "nvidia"
    },
    "type": "switch"
}

[65292] INFO: Writing requested GPU mode nvidia
[65292] INFO: Writing state {'type': 'pending_pre_xorg_start', 'requested_mode': 'nvidia', 'current_mode': 'intel'}
[516907] INFO: Received command : {
    "args": {
        "mode": "intel"
    },
    "type": "switch"
}

[516907] INFO: Writing requested GPU mode intel
[516907] INFO: Writing state {'type': 'pending_pre_xorg_start', 'requested_mode': 'intel', 'current_mode': 'nvidia'}
[731212] INFO: Received command : {
    "args": {
        "mode": "nvidia"
    },
    "type": "switch"
}

[731212] INFO: Writing requested GPU mode nvidia
[731212] INFO: Writing state {'type': 'pending_pre_xorg_start', 'requested_mode': 'nvidia', 'current_mode': 'intel'}
[816244] INFO: Received command : {
    "args": {
        "mode": "hybrid"
    },
    "type": "switch"
}

[816244] INFO: Writing requested GPU mode hybrid
[816245] INFO: Writing state {'type': 'pending_pre_xorg_start', 'requested_mode': 'hybrid', 'current_mode': 'nvidia'}
[982621] INFO: Received command : {
    "args": {
        "mode": "intel"
    },
    "type": "switch"
}

[982621] INFO: Writing requested GPU mode intel
[982621] INFO: Writing state {'type': 'pending_pre_xorg_start', 'requested_mode': 'intel', 'current_mode': 'hybrid'}

```

**Comment by Alarg** [comment]
/var/log/switch logs:
Intel -> Nvidia:
```
[6] INFO: # Xorg pre-start hook
[6] INFO: Previous state was: {'type': 'pending_pre_xorg_start', 'requested_mode': 'nvidia', 'current_mode': 'intel'}
[6] INFO: Requested mode is: nvidia
[6] INFO: Checking for GDM display servers
[200] INFO: Available modules: ['nouveau', 'bbswitch', 'acpi_call', 'nvidia', 'nvidia_drm', 'nvidia_modeset', 'nvidia_uvm']
[200] INFO: Unloading modules ['nouveau'] (if loaded)
[203] INFO: Loading module bbswitch
[205] INFO: Setting GPU power to ON via bbswitch
[361] INFO: Setting Nvidia PCI power state to on
[401] INFO: Loading module nvidia
[618] INFO: Loading module nvidia_drm
[1288] INFO: Loaded extra Intel Xorg options (0 lines)
[1288] INFO: Loaded extra Intel Xorg options (0 lines)
[1289] INFO: Writing to /etc/X11/xorg.conf.d/10-optimus-manager.conf
[1289] INFO: Writing state {'type': 'pending_post_xorg_start', 'switch_id': '20200713T141750', 'requested_mode': 'nvidia'}
[1289] INFO: Xorg pre-start hook completed successfully.
[3] INFO: # Xorg post-start hook
[3] INFO: Running xrandr commands
[269] INFO: Running /etc/optimus-manager/xsetup-nvidia.sh
[280] INFO: Writing state {'type': 'done', 'switch_id': '20200713T141750', 'current_mode': 'nvidia'}
[280] INFO: Xorg post-start hook completed successfully.
```
Nvidia -> Hybrid:
```
[6] INFO: # Xorg pre-start hook
[6] INFO: Previous state was: {'type': 'pending_pre_xorg_start', 'requested_mode': 'hybrid', 'current_mode': 'nvidia'}
[6] INFO: Requested mode is: hybrid
[6] INFO: Checking for GDM display servers
[31] INFO: Loaded extra Intel Xorg options (0 lines)
[31] INFO: Loaded extra Intel Xorg options (0 lines)
[32] INFO: Writing to /etc/X11/xorg.conf.d/10-optimus-manager.conf
[32] INFO: Writing state {'type': 'pending_post_xorg_start', 'switch_id': '20200713T141915', 'requested_mode': 'hybrid'}
[32] INFO: Xorg pre-start hook completed successfully.
[3] INFO: # Xorg post-start hook
[3] INFO: Running /etc/optimus-manager/xsetup-hybrid.sh
[18] INFO: Writing state {'type': 'done', 'switch_id': '20200713T141915', 'current_mode': 'hybrid'}
[19] INFO: Xorg post-start hook completed successfully.
```
Hybrid -> Intel:
```
[6] INFO: # Xorg pre-start hook
[6] INFO: Previous state was: {'type': 'pending_pre_xorg_start', 'requested_mode': 'intel', 'current_mode': 'hybrid'}
[6] INFO: Requested mode is: intel
[6] INFO: Checking for GDM display servers
[193] INFO: Available modules: ['nouveau', 'bbswitch', 'acpi_call', 'nvidia', 'nvidia_drm', 'nvidia_modeset', 'nvidia_uvm']
[193] INFO: Unloading modules ['nvidia_drm', 'nvidia_modeset', 'nvidia_uvm', 'nvidia'] (if loaded)
[664] INFO: Loading module bbswitch
[668] INFO: Setting GPU power to OFF via bbswitch
[741] WARNING: bbswitch is enabled, pci_power_control option ignored.
[766] INFO: Loaded extra Intel Xorg options (0 lines)
[767] INFO: Loaded extra Intel Xorg options (0 lines)
[767] INFO: Writing to /etc/X11/xorg.conf.d/10-optimus-manager.conf
[767] INFO: Writing state {'type': 'pending_post_xorg_start', 'switch_id': '20200713T142201', 'requested_mode': 'intel'}
[767] INFO: Xorg pre-start hook completed successfully.
[3] INFO: # Xorg post-start hook
[3] INFO: Running /etc/optimus-manager/xsetup-intel.sh
[10] INFO: Writing state {'type': 'done', 'switch_id': '20200713T142201', 'current_mode': 'intel'}
[10] INFO: Xorg post-start hook completed successfully.
```

**Comment by Alarg** [comment]
Journalctl from current boot:
[log.txt](https://github.com/Askannz/optimus-manager/files/4912421/log.txt)

**Comment by Alarg** [comment]
Content of 10-optimus-manager.conf after Nvidia switch:
```
Section "Files"
	ModulePath "/usr/lib/nvidia"
	ModulePath "/usr/lib32/nvidia"
	ModulePath "/usr/lib32/nvidia/xorg/modules"
	ModulePath "/usr/lib32/xorg/modules"
	ModulePath "/usr/lib64/nvidia/xorg/modules"
	ModulePath "/usr/lib64/nvidia/xorg"
	ModulePath "/usr/lib64/xorg/modules"
EndSection

Section "ServerLayout"
	Identifier "layout"
	Screen 0 "nvidia"
	Inactive "intel"
EndSection

Section "Device"
	Identifier "nvidia"
	Driver "nvidia"
	BusID "PCI:1:0:0"
	Option "Coolbits" "28"
EndSection

Section "Screen"
	Identifier "nvidia"
	Device "nvidia"
	Option "AllowEmptyInitialConfiguration"
EndSection

Section "Device"
	Identifier "intel"
	Driver "modesetting"
	BusID "PCI:0:2:0"
EndSection

Section "Screen"
	Identifier "intel"
	Device "intel"
EndSection

Section "ServerFlags"
	Option "IgnoreABI" "1"
EndSection
```

**Comment by Hyper-KVM** [comment]
I think you don't understand how PRIME works, you see, since the iGPU in most cases is connected to the display, with the dGPU being connected to the iGPU and not the display, the iGPU basically acts as a frame buffer for the dGPU to pass it's output to. Which means that the iGPU is required to be on to output anything to the screen.

**Comment by Alarg** [comment]
> I think you don't understand how PRIME works, you see, since the iGPU in most cases is connected to the display, with the dGPU being connected to the iGPU and not the display, the iGPU basically acts as a frame buffer for the dGPU to pass it's output to. Which means that the iGPU is required to be on to output anything to the screen. As far as I can tell, everything is working correctly from the logs.

I understand how it works. The issue is, on my previous system install, I did everything the way it was mentioned on the git page + sddm.conf change
On the previous install, running `optimus-manager --nvidia` would give a result like:
```
[alarg@ALEKSA ~]$ xrandr --listproviders
Providers: number : 1
Provider 0: id: 0x44 cap: 0xf, Source Output, Sink Output, Source Offload, Sink Offload crtcs: 3 outputs: 2 associated providers: 0 name:NVIDIA-G0
```

That's the issue. Look again at `xrandr --listproviders` outputs (first post) and look who's sink and who's provider. Before the reinstall, I'd get different outputs from Hybrid modes. I can't run apps in Nvidia mode (e.g. Minecraft), and can't start with Prime in Hybrid mode.
[Also this](https://download.nvidia.com/XFree86/Linux-x86_64/440.64/README/primerenderoffload.html).

I'm not quite sure if this is an issue for Optimus-manager or not, anymore. I see that Xorg and Nvidia drivers might be to blame (I'm using Nvidia hybrid drivers by the way).

**Comment by Alarg** [comment]
I tried several different kernels, and it stil doesn't work (along with different bbswitch packages). I can only assume now that it must be either a drivers' issue or the manager's issue. I can't take blind guesses, so I hope that Asskanz might help out soon enough. It's certain that an update of some package broke this.
I don't know Python, unfortunately, and I don't have much time to investigate currently. I might later, but for now, if Asskanz notices, I hope we can get to the bottom of this.
Mind you, again, the previous installation of Manjaro (same ISO) worked perfectly.

I also now get default video-linux driver installed instead of video-modesetting I got on the first install, which is weird.
